### PR TITLE
Feature/database

### PR DESCRIPTION
**Pull request by norling**
This PR includes only the code database functions that are currently in [sda-pipeline](https://github.com/neicnordic/sda-pipeline/tree/master/internal/database). 
Moving the code out of sda-pipeline is to allow the same codebase to be used in non-pipeline services. 

This PR is intended to evaluate the base structure of the database code, as well as the test coverage of the core functions. The functions that does queries used by services will be added in a separate PR to keep PR size down.

### CODE REVIEW

**Comment by jbygdell** [comment]
So question number 1: Why did you start from scratch instead of just take all there is in the `sda-pipeline` repo?
That way the transition would have been seamless.

Question number 2: why does a PR contain `merge` commits?

**Comment by norling** [comment]
It's taken from sda-pipeline, but shifted around to be (according to me) a bit cleaner and better commented. :) And making the transition seamless didn't seem like a priority to me compared to having more readable code. There will need to be updates in the services to account for the v4 schema anyway, and there's no great rush to change until then...
And it's got merge commits because there was a PR into this branch with the github actions, and I don't care about cleaning history! :D

**Review by dbampalikis — APPROVED** [review]

File: database/db_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	// check that we can't do queries on a closed connection
+	query := "SELECT MAX(version) FROM local_ega.dbschema_version"
+	var dbVersion = -1
+	err = db.db.QueryRow(query).Scan(&dbVersion)
+	assert.NotNil(suite.T(), err, "query possible on closed connection")
+
+	// check that nothing happens if Close is called on a closed connection
+	for i := 0; i < 10; i++ {
+		db.Close()
+	}
```
Comment (aaperis) [inline database/db_test.go]: I see the pragmatic aspect of the test. However, I would prefer it if the error is documented in the test's output somehow.

**Review by aaperis — APPROVED** [review]
There's a small issue with one of the tests which I would have liked to be handled differently but I can live with that. Overall it looks nice and clean :-).

**Comment by blankdots** [comment]
does it make sense to include functions necessary for the download part as well e.g. https://github.com/neicnordic/sda-download/blob/main/internal/database/database.go ?

**Review by blankdots — APPROVED** [review]

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+on: [push]
+
+jobs:
+
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        go-version: [1.16, 1.17, 1.18]
```
Comment (blankdots) [inline .github/workflows/test.yml]:
```suggestion
        go-version: [1.18, 1.19]
```
might want to address these

**Pull request merged**